### PR TITLE
AE-1214: Low Liquidity Market Attack Monitor

### PR DESCRIPTION
**Pull request by arbitrary-CodeBeholder**

### CODE REVIEW

File: low-liquidity-market-attack-monitor/package.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "forta-ae-market-attack-events",
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/package.json:2]:
```suggestion
  "name": "forta-ae-comp-market-attack-events",
```

File: low-liquidity-market-attack-monitor/src/agent.js [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  compTokenAbi,
+  excludeAddresses,
+  compTokens,
+) {
+  let compTokenAddresses = await comptrollerContract.getAllMarkets();
+  compTokenAddresses = compTokenAddresses.map((addr) => addr.toLowerCase());
+  compTokenAddresses = compTokenAddresses.filter((addr) => excludeAddresses.indexOf(addr) === -1);
+  compTokenAddresses = compTokenAddresses.filter(
+    (addr) => !Object.keys(compTokens).includes(addr),
+  );
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.js:59]: Same functionality, just using chaining in this suggestion.
```suggestion
  compTokenAddresses = compTokenAddresses
    .map((addr) => addr.toLowerCase())
    .filter((addr) => excludeAddresses.indexOf(addr) === -1)
    .filter((addr) => !Object.keys(compTokens).includes(addr));
```

File: low-liquidity-market-attack-monitor/src/agent.js [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    data.excludeAddresses = excludeAddresses.map((addr) => addr.toLowerCase());
+
+    data.provider = getEthersBatchProvider();
+
+    const {
+      Comptroller: comptroller,
+      CompoundToken: compToken,
+    } = config.contracts;
+
+    // from the Comptroller contract, get all of the cTokens
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.js:88]:
```suggestion
    // create an ethers.js Contract Object to interact with the Comptroller contract
```

File: low-liquidity-market-attack-monitor/src/agent.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    data.comptrollerContract = new ethers.Contract(
+      comptroller.address,
+      comptrollerAbi,
+      data.provider,
+    );
+
+    // cToken contracts
+    data.compTokenAbi = getAbi(compToken.abiFile);
+    data.compTokens = {};
+
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.js:99]:
```suggestion

    // from the Comptroller contract, get all of the cTokens
```

File: low-liquidity-market-attack-monitor/src/agent.js [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+          developerAbbreviation,
+          compTokens[transferEvent.args.to],
+          transferEvent.args.to,
+          mintEvent.args.mintAmount.toString(),
+          mintEvent.args.mintTokens.toString(),
+          transferEvent.args.from,
+          transferEvent.args.amount.toString(),
+        ));
+      });
+    });
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.js:152]: I don't think this is checking for the correct condition.

It looks like it is checking for the following:
- Determine if an ERC20 Transfer event occurred with any of the cToken addresses as the recipient (`to` field)
- If so, get any Mint events in the transaction that were emitted from the cToken address that was the recipient of the Transfer event

For this potential condition, I think we need to check for the following:
- Did mint() get called on any cToken contract?
  - A successful mint() call will result in a Mint event immediately followed by a Transfer event
- Then, in the same transaction, was the `underlying` token then directly transferred to that same cToken contract, which should be indicated by another Transfer event (this one won't have a Mint event right before it)?

So we need to check that a Transfer event to a cToken was emitted from the `underlying` token address (corresponding to that specific cToken).
We also need to check the ordering of the events in the log.  If the `underlying` token is transferred to the cToken contract _before_ the mint() call, that would be a botched attempt to take advantage of this issue.  That is to say, the order of the events should be:
- Mint emitted by the cToken contract
- Transfer emitted by the cToken contract
- Transfer emitted by the `underlying` token contract with the cToken address as the recipient

File: low-liquidity-market-attack-monitor/src/agent.spec.js [hunk 6]
```diff
@@ -0,0 +1,380 @@
+const mockedGetContract = jest.fn();
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.spec.js:1]: These tests need to be updated to work with the newer version of the Forta SDK.

I ran this locally and the tests that used the `mockTxEvent` Object failed:
```bash
rm -rf node_modules package-lock.json
npm i
npm test
```

The arguments for the `TransactionEvent` constructor have changed: https://github.com/forta-protocol/forta-agent-sdk/blob/d46e9a7e1f547296cc85c23fae7c22e8c3ce3fbf/sdk/transaction.event.ts#L19

Here is an example of us using the newer argument ordering:
https://github.com/arbitraryexecution/forta-agent-templates/blob/be74f81f86f5dbceaa80e4b4a155ed1f3bde1a31/gnosis-safe-multisig/src/agent.spec.js#L189

The `receipt` key no longer exists, but the `logs` key has been promoted to be directly under the `txEvent` Object.

File: low-liquidity-market-attack-monitor/agent-config.json [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+      "abiFile": "Comptroller.json"
+    },
+    "CompoundToken": {
+      "address": "0xc00e94cb662c3520282e6f5717214004a7f26888",
+      "abiFile": "cErc20.json"
+    }
+  },
+  "excludeAddresses": [
+    "0x4ddc2d193948926d02f9b1fe9e1daa0718270ed5",
+    "0x70e36f6bf80a52b3b46b3af8e106cc0ed743e8e4"
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/agent-config.json:18]: Why are we excluding the `cCOMP` token?

File: low-liquidity-market-attack-monitor/src/agent.js [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    const symbol = await contract.symbol();
+    const underlying = await contract.underlying();
+
+    // eslint-disable-next-line no-param-reassign
+    compTokens[tokenAddress] = {
+      symbol,
+      underlying,
+    };
+  });
+  await Promise.all(promises);
```
Comment (Nomad-AE) [inline low-liquidity-market-attack-monitor/src/agent.js:101]: ``
```suggestion
  await Promise.all(compTokenAddresses.map(async (tokenAddress) => {
    const contract = new ethers.Contract(tokenAddress, compTokenAbi, provider);
    const symbol = await contract.symbol();
    const underlying = await contract.underlying();

    // eslint-disable-next-line no-param-reassign
    compTokens[tokenAddress] = {
      symbol,
      underlying,
    };
  }));
```
If you aren't doing any other async calls, then you can simplify a little bit right here. You should double check with @echoAE if my recommendation is OK or should we stick with our original patterns.

**Review by Nomad-AE — APPROVED** [review]
Code looks good and all tests pass.

File: low-liquidity-market-attack-monitor/src/agent.js [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+    );
+
+    const transferEvents = filterLog(txEvent.logs, ERC20_TRANSFER_EVENT);
+
+    transferEvents.forEach((transferEvent) => {
+      Object.entries(compTokens).forEach(([compTokenAddress, compToken]) => {
+        if (transferEvent.address === compToken.underlying
+          && transferEvent.args.to === compTokenAddress) {
+          filterLog(txEvent.logs, CERC20_MINT_EVENT)
+            .filter((mintEvent) => mintEvent.address === compTokenAddress)
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.js:173]: I cringe at even the thought of mentioning this (really, couldn't we just agree to used checksummed addresses everywhere?), but should we use `.toLowerCase()` for all of the address comparisons to keep the code consistent?
```suggestion
        if (transferEvent.address.toLowerCase() === compToken.underlying.toLowerCase()
          && transferEvent.args.to.toLowerCase() === compTokenAddress.toLowerCase()) {
          filterLog(txEvent.logs, CERC20_MINT_EVENT)
            .filter((mintEvent) => mintEvent.address.toLowerCase() === compTokenAddress.toLowerCase())
```

Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.js:173]: The `filterLog` function in this module has the ability to filter based on log address, so these two lines can probably be combined:
```suggestion
          filterLog(txEvent.logs, CERC20_MINT_EVENT, compTokenAddress)
```

File: low-liquidity-market-attack-monitor/src/agent.js [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+            // This shouldn't really ever return more than one result not sure
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.js:174]: Can you remove this comment and place a single comment above this entire block of code that contains the filtering logic?

Something to the effect of:
```javascript
// return findings under the following circumstances
//   - A Mint event was emitted before a Transfer event
//   - The Mint event was emitted by a Compound cToken contract
//   - The Transfer event was emitted by the underlying token contract corresponding to that Compound cToken contract
//   - The Transfer event was for the underlying token being transferred directly to the Compound cToken contract
```

File: low-liquidity-market-attack-monitor/agent-config.json [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+  "developerAbbreviation": "AE",
+  "protocolName": "Compound",
+  "protocolAbbreviation": "COMP",
+  "contracts": {
+    "Comptroller": {
+      "address": "0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B",
+      "abiFile": "Comptroller.json"
+    },
+    "CompoundToken": {
+      "address": "0xc00e94cb662c3520282e6f5717214004a7f26888",
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/agent-config.json:12]: This does not appear to be used anywhere in a substantive way.  It's also a little confusing to have the main key called `CompoundToken`, but that's not really what this entry is used for in the code (i.e. the COMP token isn't involved in this Bot).

File: low-liquidity-market-attack-monitor/src/agent.spec.js [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+
+    const { Comptroller, CompoundToken } = contracts;
+    expect(typeof Comptroller).toBe('object');
+    expect(Comptroller).not.toBe({});
+
+    expect(typeof CompoundToken).toBe('object');
+    expect(CompoundToken).not.toBe({});
+
+    const { abiFile: ComptrollerAbiFile, address: ComptrollerAddress } = Comptroller;
+    const { abiFile: cTokenAbiFile, address: compTokenAddress } = CompoundToken;
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.spec.js:77]:
```suggestion
    const { abiFile: cTokenAbiFile, address: compTokenAddress } = CompoundToken;
```

File: low-liquidity-market-attack-monitor/src/agent.spec.js [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+
+    // check that the comptroller address is a valid address
+    expect(ethers.utils.isHexString(ComptrollerAddress, 20)).toBe(true);
+
+    // check that the compound token address is a valid address
+    expect(ethers.utils.isHexString(compTokenAddress, 20)).toBe(true);
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.spec.js:83]:
```suggestion
```

File: low-liquidity-market-attack-monitor/src/agent.spec.js [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+    let validCompTokenAddress = `0x1${'0'.repeat(39)}`;
+    let validCompTokenSymbol = 'TEST';
+    let validAttackAddress = `0x9${'0'.repeat(39)}`;
+    /* eslint-enable prefer-const */
+
+    beforeEach(async () => {
+      initializeData = {};
+
+      mockComptrollerContract = {
+        getAllMarkets: jest.fn().mockReturnValueOnce([validCompTokenAddress]),
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.spec.js:183]:
```suggestion
        getAllMarkets: jest.fn().mockResolvedValueOnce([validCompTokenAddress]),
```

File: low-liquidity-market-attack-monitor/src/agent.spec.js [hunk 15]
```diff
@@ -0,0 +1,387 @@
...
+      };
+
+      mockedGetContract.mockReturnValueOnce(mockComptrollerContract);
+
+      mockedCompTokenContract = {
+        symbol: jest.fn().mockReturnValueOnce(validCompTokenSymbol),
+        underlying: jest.fn().mockReturnValueOnce(validUnderlyingAddress),
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.spec.js:190]:
```suggestion
        symbol: jest.fn().mockResolvedValueOnce(validCompTokenSymbol),
        underlying: jest.fn().mockResolvedValueOnce(validUnderlyingAddress),
```

File: low-liquidity-market-attack-monitor/src/agent.spec.js [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+      compTokenAbi = initializeData.compTokenAbi;
+      compTokenInterface = new ethers.utils.Interface(compTokenAbi);
+
+      mockTxEvent = createTransactionEvent({
+        logs: [],
+      });
+    });
+
+    it('returns empty findings if no direct transfer events were emitted in the transaction', async () => {
+      mockComptrollerContract.getAllMarkets.mockReturnValueOnce([]);
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.spec.js:212]:
```suggestion
      mockComptrollerContract.getAllMarkets.mockResolvedValueOnce([]);
```

File: low-liquidity-market-attack-monitor/src/agent.spec.js [hunk 17]
```diff
@@ -0,0 +1,387 @@
...
+
+      const findings = await handleTransaction(mockTxEvent);
+
+      expect(findings).toStrictEqual([]);
+    });
+
+    it('returns empty findings if no mint events were emitted in the transaction', async () => {
+      mockComptrollerContract.getAllMarkets.mockReturnValueOnce([]);
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.spec.js:220]:
```suggestion
      mockComptrollerContract.getAllMarkets.mockResolvedValueOnce([]);
```

File: low-liquidity-market-attack-monitor/src/agent.spec.js [hunk 18]
```diff
@@ -0,0 +1,387 @@
...
+      const transferEventAbi = compTokenInterface.getEvent('Transfer');
+      const transferEvent = createMockEventLogs(transferEventAbi, compTokenInterface, override);
+
+      const transferLog = {
+        address: validAttackAddress,
+        topics: transferEvent.mockTopics,
+        args: transferEvent.mockArgs,
+        data: transferEvent.data,
+        signature: transferEventAbi.format(ethers.utils.FormatTypes.minimal).substring(6),
+      };
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.spec.js:237]: I don't believe all of these fields are necessary (we included additional fields in previous tests that did not need to be mocked to create a TransactionEvent):
```suggestion
      const transferLog = {
        address: validAttackAddress,
        topics: transferEvent.mockTopics,
        data: transferEvent.data,
      };
```

File: low-liquidity-market-attack-monitor/src/agent.spec.js [hunk 19]
```diff
@@ -0,0 +1,387 @@
...
+      const transferEvent = createMockEventLogs(transferEventAbi, compTokenInterface, override);
+
+      const transferLog = {
+        logIndex: 10,
+        address: validUnderlyingAddress,
+        topics: transferEvent.mockTopics,
+        args: transferEvent.mockArgs,
+        data: transferEvent.data,
+        signature: transferEventAbi.format(ethers.utils.FormatTypes.minimal).substring(6),
+      };
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.spec.js:267]:
```suggestion
      const transferLog = {
        logIndex: 10,
        address: validUnderlyingAddress,
        topics: transferEvent.mockTopics,
        data: transferEvent.data,
      };
```

File: low-liquidity-market-attack-monitor/src/agent.spec.js [hunk 20]
```diff
@@ -0,0 +1,387 @@
...
+
+      mockTxEvent.logs.push(transferLog);
+
+      const findings = await handleTransaction(mockTxEvent);
+
+      expect(findings).toStrictEqual([]);
+    });
+
+    it('returns finding if attacker minted cTokens and made a direct transfer in the transaction', async () => {
+      mockComptrollerContract.getAllMarkets.mockReturnValueOnce([]);
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.spec.js:247]:
```suggestion
      mockComptrollerContract.getAllMarkets.mockResolvedValueOnce([]);
```

File: low-liquidity-market-attack-monitor/src/agent.spec.js [hunk 21]
```diff
@@ -0,0 +1,387 @@
...
+      const mintEvent = createMockEventLogs(mintEventAbi, compTokenInterface, override);
+
+      const mintLog = {
+        logIndex: 5,
+        address: validCompTokenAddress,
+        topics: mintEvent.mockTopics,
+        args: mintEvent.mockArgs,
+        data: mintEvent.data,
+        signature: mintEventAbi.format(ethers.utils.FormatTypes.minimal).substring(6),
+      };
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.spec.js:290]:
```suggestion
      const mintLog = {
        logIndex: 5,
        address: validCompTokenAddress,
        topics: mintEvent.mockTopics,
        data: mintEvent.data,
      };
```

File: low-liquidity-market-attack-monitor/src/agent.spec.js [hunk 22]
```diff
@@ -0,0 +1,387 @@
...
+
+      expect(findings).toStrictEqual([expectedFinding]);
+    });
+
+    it('returns finding if new market was added and attacked', async () => {
+      const newUnderlyingAddress = `0x6${'0'.repeat(39)}`;
+      const newCompTokenAddress = `0x2${'0'.repeat(39)}`;
+      const newCompSymbol = 'NEWTEST';
+
+      mockComptrollerContract.getAllMarkets.mockReturnValueOnce([newCompTokenAddress]);
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.spec.js:316]:
```suggestion
      mockComptrollerContract.getAllMarkets.mockResolvedValueOnce([newCompTokenAddress]);
```

File: low-liquidity-market-attack-monitor/src/agent.spec.js [hunk 23]
```diff
@@ -0,0 +1,387 @@
...
+
+      mockedCompTokenContract = {
+        symbol: jest.fn().mockReturnValueOnce(newCompSymbol),
+        underlying: jest.fn().mockReturnValueOnce(newUnderlyingAddress),
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.spec.js:320]:
```suggestion
        symbol: jest.fn().mockResolvedValueOnce(newCompSymbol),
        underlying: jest.fn().mockResolvedValueOnce(newUnderlyingAddress),
```

File: low-liquidity-market-attack-monitor/src/agent.spec.js [hunk 24]
```diff
@@ -0,0 +1,387 @@
...
+      const transferEvent = createMockEventLogs(transferEventAbi, compTokenInterface, override);
+
+      const transferLog = {
+        logIndex: 10,
+        address: newUnderlyingAddress,
+        topics: transferEvent.mockTopics,
+        args: transferEvent.mockArgs,
+        data: transferEvent.data,
+        signature: transferEventAbi.format(ethers.utils.FormatTypes.minimal).substring(6),
+      };
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.spec.js:343]:
```suggestion
      const transferLog = {
        logIndex: 10,
        address: newUnderlyingAddress,
        topics: transferEvent.mockTopics,
        data: transferEvent.data,
      };
```

File: low-liquidity-market-attack-monitor/src/agent.spec.js [hunk 25]
```diff
@@ -0,0 +1,387 @@
...
+      const mintEvent = createMockEventLogs(mintEventAbi, compTokenInterface, override);
+
+      const mintLog = {
+        logIndex: 5,
+        address: newCompTokenAddress,
+        topics: mintEvent.mockTopics,
+        args: mintEvent.mockArgs,
+        data: mintEvent.data,
+        signature: mintEventAbi.format(ethers.utils.FormatTypes.minimal).substring(6),
+      };
```
Comment (echoAE) [inline low-liquidity-market-attack-monitor/src/agent.spec.js:366]:
```suggestion
      const mintLog = {
        logIndex: 5,
        address: newCompTokenAddress,
        topics: mintEvent.mockTopics,
        data: mintEvent.data,
      };
```

**Review by echoAE — APPROVED** [review]
A few changes suggested, but none that should change the functionality of the code.

File: low-liquidity-market-attack-monitor/src/agent.js (hunk 9 above)
Comment (arbitrary-CodeBeholder) [reply low-liquidity-market-attack-monitor/src/agent.js:173]: The Forta SDK does create the event with the address already lowered, but yes we should convert the return value from any contract call such as `underlying`

File: low-liquidity-market-attack-monitor/agent-config.json (hunk 11 above)
Comment (arbitrary-CodeBeholder) [reply low-liquidity-market-attack-monitor/agent-config.json:12]: Yea it's a bad naming collision... cToken vs Compound Token

**Pull request merged**